### PR TITLE
Feature/9280 scan qr codes from image file/9410 display errors

### PR DESCRIPTION
**Pull request by KaiTeuber**
## Description
Show errors for noQRCode, wrongQRCode or invalid file

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-9410

## Video
(on a slow device)
https://user-images.githubusercontent.com/2675578/135491256-79f0e4be-ad87-42ce-8bc0-176b1569c568.mp4

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 1]
```diff
@@ -27,12 +28,14 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
...
 		self.missingPasswordForPDF = missingPasswordForPDF
 		self.failedToUnlockPDF = failedToUnlockPDF
+		self.presentSimpleAlert = presentSimpleAlert
 	}
 
 	// MARK: - Protocol PHPickerViewControllerDelegate
 
 	@available(iOS 14, *)
 	func picker(_ picker: PHPickerViewController, didFinishPicking results: [PHPickerResult]) {
+		dismiss()
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:38]:
```suggestion
		didFinishPicking()
```

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 2]
```diff
@@ -18,7 +18,8 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
 		qrCodeFound: @escaping (QRCodeResult?) -> Void,
 		qrCodeParser: QRCodeParsable,
 		missingPasswordForPDF: @escaping (@escaping (String) -> Void) -> Void,
-		failedToUnlockPDF: @escaping () -> Void
+		failedToUnlockPDF: @escaping () -> Void,
+		presentSimpleAlert: @escaping (FileScannerCoordinator.AlertTypes) -> Void
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:22]:
```suggestion
		didFailPickingFile: @escaping (FileScannerCoordinator.FilePickingError) -> Void
```

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 3]
```diff
@@ -27,12 +28,14 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
...
 		self.failedToUnlockPDF = failedToUnlockPDF
+		self.presentSimpleAlert = presentSimpleAlert
 	}
 
 	// MARK: - Protocol PHPickerViewControllerDelegate
 
 	@available(iOS 14, *)
 	func picker(_ picker: PHPickerViewController, didFinishPicking results: [PHPickerResult]) {
+		dismiss()
 		showHUD()
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:39]:
```suggestion
		processingFileStarted()
```

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 4]
```diff
@@ -52,14 +55,14 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
 					  let codes = self.findQRCodes(in: image),
 					  !codes.isEmpty
 				else {
-					self?.dismissOnMain()
+					self?.presentSimpleAlert(.noQRCodeFound)
+					self?.hideHUD()
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:59]:
```suggestion
					self?.processingFileFinished()
```

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift [hunk 5]
```diff
@@ -180,7 +185,7 @@ class FileScannerCoordinatorViewModel: NSObject, PHPickerViewControllerDelegate,
 			self.findValidQRCode(from: codes) { [weak self] result in
 				self?.qrCodeFound(result)
 				self?.hideHUD()
-				self?.dismiss()
+//				self?.dismiss()
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinatorViewModel.swift:188]: Can be removed?

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinator.swift [hunk 6]
```diff
@@ -13,12 +13,12 @@ class FileScannerCoordinator {
 		_ parentViewController: UIViewController,
 		qrCodeFound: @escaping (QRCodeResult?) -> Void,
 		qrCodeParser: QRCodeParsable,
-		dismiss: @escaping () -> Void
+		endFlow: @escaping () -> Void
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinator.swift:16]:
```suggestion
		noQRCodeFound: @escaping () -> Void
```

File: src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinator.swift [hunk 7]
```diff
@@ -44,14 +44,18 @@ class FileScannerCoordinator {
...
 			},
 			qrCodeParser: qrCodeParser,
-			missingPasswordForPDF: { callback in
-				self.presentPasswordAlert(callback)
+			missingPasswordForPDF: { [weak self] callback in
+				self?.presentPasswordAlert(callback)
 			},
-			failedToUnlockPDF: {
-				self.presentPDFUnlockFailedAlert()
+			failedToUnlockPDF: { [weak self] in
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/FileScanner/FileScannerCoordinator.swift:54]: Can we remove this callback and replace it by `presentSimpleAlert`

**Review by ArturFriesen — APPROVED** [review]
I think we should change the naming. Because the view model should not tell the coordinator "what to do", but instead it should tell the coordinator "what happened" in his context. The coordinator should then decide "what to do" on his own, based on "what happened" in the view model. 🤓

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings [hunk 8]
```diff
@@ -3215,6 +3215,22 @@ Die Angaben werden statistisch ausgewertet, sie werden nicht zu einem Profil ges
 
 "FileScanner_PasswordError_message" = "Bitte geben Sie ein gültiges Passwort ein.";
 
+"FileScanner_InvalidQRCode_title" = "Kein geeigneter QR-Code";
+
+"FileScanner_InvalidQRCode_message" = "Der QR-Code wird nicht in der Corona-Warn-App unterstützt. Bitte wählen Sie eine Datei mit einem geeigneten QR-Code.";
```
[the remark this answers is not in the document]
Comment (CV113) [reply src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:3220]: Hi Tim, thank you very much for the proposal, we will change it.

Comment (CV113) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:3220]:
```suggestion
"FileScanner_InvalidQRCode_message" = "Der QR-Code wird nicht von der Corona-Warn-App unterstützt. Bitte wählen Sie eine Datei mit einem geeigneten QR-Code."; 
```

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings [hunk 9]
```diff
@@ -3215,6 +3215,22 @@ Die Angaben werden statistisch ausgewertet, sie werden nicht zu einem Profil ges
...
+
+"FileScanner_InvalidQRCode_title" = "Kein geeigneter QR-Code";
+
+"FileScanner_InvalidQRCode_message" = "Der QR-Code wird nicht in der Corona-Warn-App unterstützt. Bitte wählen Sie eine Datei mit einem geeigneten QR-Code.";
```
Comment (CV113) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:3224]:
```suggestion
"FileScanner_InvalidQRCode_message" = "Der QR-Code wird nicht von der Corona-Warn-App unterstützt. Bitte wählen Sie eine Datei mit einem geeigneten QR-Code.";
```

**Review by CV113 — APPROVED** [review]
UA approved

**Review by marcussc — APPROVED** [review]

**Pull request merged**